File: src/main/java/org/athenian/LetterCount.java
package org.athenian;

public class LetterCount {

    public static int letterCountLoop(char c, String s) {
        String cstr = Character.toString(c);
        int cnt = 0;

        for (int i = 0; i < s.length(); i++)
            if (s.substring(i, i + 1).equals(cstr))
                cnt++;

        return cnt;
    }

    public static int letterCountRecursion(char c, String s) {
        if (s.length() == 0)
            return 0;

        String firstLetter = s.substring(0, 1);
        String remainingLetters = s.substring(1);

        if (firstLetter.equals(Character.toString(c)))
            return 1 + letterCountRecursion(c, remainingLetters);
        else
            return letterCountRecursion(c, remainingLetters);
    }

    public static int letterCountStream(char c, String s) {
        return (int) s.chars()
                .filter(value -> value == c)
                .mapToLong(value -> 1)
                .sum();
    }
}

File: src/main/kotlin/org/athenian/LetterCount.kt
package org.athenian


fun letterCountLoop(c: Char, s: String): Int {
    var cnt = 0

    for (aChar in s)
        if (aChar == c)
            cnt++

    return cnt
}

fun letterCountRecursion(c: Char, s: String): Int {
    if (s.length == 0)
        return 0

    val firstLetter = s.slice(0..0)
    val remainingLetters = s.slice(1..s.length - 1)

    return (if (firstLetter == Character.toString(c)) 1 else 0) + letterCountRecursion(c, remainingLetters)
}

fun letterCountStream(c: Char, s: String) =
    s.chars()
        .filter { it == c.toInt() }
        .map { 1 }
        .sum()